src/mlpack/methods/ann/ffn_impl.hpp

```diff
@@ -192,101 +122,254 @@ LayerTypes, OutputLayerType, InitializationRuleType, PerformanceFunction
+
+    if (!reset)
+    {
+      // Get the output width.
+      if (boost::apply_visitor(outputWidthVisitor, network[i]) != 0)
+      {
+        width = boost::apply_visitor(outputWidthVisitor, network[i]);
+      }
+
+      // Get the output height.
+      if (boost::apply_visitor(outputHeightVisitor, network[i]) != 0)
+      {
+        height = boost::apply_visitor(outputHeightVisitor, network[i]);
+      }
+    }
+  }
+
+  if (!reset)
+  {
+    reset = true;
+  }
+}
+
+template<typename OutputLayerType, typename InitializationRuleType>
+void FFN<OutputLayerType, InitializationRuleType>::Backward()
+{
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, network.back())), std::move(error), std::move(
+      boost::apply_visitor(deltaVisitor, network.back()))), network.back());
+
+  for (size_t i = 2; i < network.size(); ++i)
+  {
+    boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+        outputParameterVisitor, network[network.size() - i])), std::move(
+        boost::apply_visitor(deltaVisitor, network[network.size() - i + 1])),
+        std::move(boost::apply_visitor(deltaVisitor,
+        network[network.size() - i]))), network[network.size() - i]);
+  }
+}
+
+template<typename OutputLayerType, typename InitializationRuleType>
+void FFN<OutputLayerType, InitializationRuleType>::Gradient()
+{
+  boost::apply_visitor(GradientVisitor(std::move(currentInput), std::move(
+      boost::apply_visitor(deltaVisitor, network[1]))), network.front());
+
+  for (size_t i = 1; i < network.size() - 1; ++i)
+  {
+    boost::apply_visitor(GradientVisitor(std::move(boost::apply_visitor(
+        outputParameterVisitor, network[i - 1])), std::move(
+        boost::apply_visitor(deltaVisitor, network[i + 1]))), network[i]);
+  }
+
+  boost::apply_visitor(GradientVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, network[network.size() - 2])), std::move(error)),
+      network[network.size() - 1]);
+}
+
+template<typename OutputLayerType, typename InitializationRuleType>
 template<typename Archive>
-void FFN<
-LayerTypes, OutputLayerType, InitializationRuleType, PerformanceFunction
->::Serialize(Archive& ar, const unsigned int /* version */)
+void FFN<OutputLayerType, InitializationRuleType>::Serialize(
+    Archive& ar, const unsigned int /* version */)
 {
   ar & data::CreateNVP(parameter, "parameter");
+  ar & data::CreateNVP(width, "width");
+  ar & data::CreateNVP(height, "height");
 
   // If we are loading, we need to initialize the weights.
   if (Archive::is_loading::value)
   {
-    NetworkWeights(parameter, network);
+    reset = false;
+
+    size_t offset = 0;
+    for (size_t i = 0; i < network.size(); ++i)
+    {
+      offset += boost::apply_visitor(WeightSetVisitor(std::move(parameter),
+          offset), network[i]);
+
+      boost::apply_visitor(resetVisitor, network[i]);
+    }
```

Comment:
If somebody calls the Gradient function right after loading the model before calling the ``Evaluate`` or ``Train`` function right.